### ISSUE TITLE
jnp fails after a = np.zeros((100, 100), dtype=np.int4) on mac m1

### ISSUE THREAD

**Issue by Maxon081102** [google/jax#21883]
### Description

```python
import jax
import jax.numpy as np
a = np.zeros((100, 100), dtype=np.int4)
```
```
---------------------------------------------------------------------------
XlaRuntimeError                           Traceback (most recent call last)
Cell In[4], line 3
      1 import jax
      2 import jax.numpy as np
----> 3 a = np.zeros((100, 100), dtype=np.int4)

File ~/Documents/miniconda3/envs/dl/lib/python3.10/site-packages/jax/_src/numpy/lax_numpy.py:2782, in zeros(shape, dtype, device)
   2780 dtypes.check_user_dtype_supported(dtype, "zeros")
   2781 shape = canonicalize_shape(shape)
-> 2782 return lax.full(shape, 0, _jnp_dtype(dtype), sharding=_normalize_to_sharding(device))

File ~/Documents/miniconda3/envs/dl/lib/python3.10/site-packages/jax/_src/lax/lax.py:1257, in full(shape, fill_value, dtype, sharding)
   1254   shard = broadcast(fill_value, broadcast_shape)
   1255   return array.make_array_from_callback(shape, sharding, lambda _: shard)
-> 1257 return broadcast(fill_value, shape)

File ~/Documents/miniconda3/envs/dl/lib/python3.10/site-packages/jax/_src/lax/lax.py:802, in broadcast(operand, sizes)
    788 """Broadcasts an array, adding new leading dimensions
    789 
    790 Args:
   (...)
    799   jax.lax.broadcast_in_dim : add new dimensions at any location in the array shape.
    800 """
    801 dims = tuple(range(len(sizes), len(sizes) + np.ndim(operand)))
--> 802 return broadcast_in_dim(operand, tuple(sizes) + np.shape(operand), dims)

File ~/Documents/miniconda3/envs/dl/lib/python3.10/site-packages/jax/_src/lax/lax.py:831, in broadcast_in_dim(operand, shape, broadcast_dimensions)
    829 else:
    830   dyn_shape, static_shape = [], shape  # type: ignore
--> 831 return broadcast_in_dim_p.bind(
    832     operand, *dyn_shape, shape=tuple(static_shape),
    833     broadcast_dimensions=tuple(broadcast_dimensions))

File ~/Documents/miniconda3/envs/dl/lib/python3.10/site-packages/jax/_src/core.py:387, in Primitive.bind(self, *args, **params)
    384 def bind(self, *args, **params):
    385   assert (not config.enable_checks.value or
    386           all(isinstance(arg, Tracer) or valid_jaxtype(arg) for arg in args)), args
--> 387   return self.bind_with_trace(find_top_trace(args), args, params)

File ~/Documents/miniconda3/envs/dl/lib/python3.10/site-packages/jax/_src/core.py:391, in Primitive.bind_with_trace(self, trace, args, params)
    389 def bind_with_trace(self, trace, args, params):
    390   with pop_level(trace.level):
--> 391     out = trace.process_primitive(self, map(trace.full_raise, args), params)
    392   return map(full_lower, out) if self.multiple_results else full_lower(out)

File ~/Documents/miniconda3/envs/dl/lib/python3.10/site-packages/jax/_src/core.py:879, in EvalTrace.process_primitive(self, primitive, tracers, params)
    877   return call_impl_with_key_reuse_checks(primitive, primitive.impl, *tracers, **params)
    878 else:
--> 879   return primitive.impl(*tracers, **params)

File ~/Documents/miniconda3/envs/dl/lib/python3.10/site-packages/jax/_src/dispatch.py:86, in apply_primitive(prim, *args, **params)
     84 prev = lib.jax_jit.swap_thread_local_state_disable_jit(False)
     85 try:
---> 86   outs = fun(*args)
     87 finally:
     88   lib.jax_jit.swap_thread_local_state_disable_jit(prev)

    [... skipping hidden 15 frame]

File ~/Documents/miniconda3/envs/dl/lib/python3.10/site-packages/jax/_src/compiler.py:237, in backend_compile(backend, module, options, host_callbacks)
    232   return backend.compile(built_c, compile_options=options,
    233                          host_callbacks=host_callbacks)
    234 # Some backends don't have `host_callbacks` option yet
    235 # TODO(sharadmv): remove this fallback when all backends allow `compile`
    236 # to take in `host_callbacks`
--> 237 return backend.compile(built_c, compile_options=options)

XlaRuntimeError: INVALID_ARGUMENT: during context [hlo verifier]: s4 is currently only supported in convert instructions, but got instruction: %broadcast.2 = s4[100,100]{1,0} broadcast(s4[] %Arg_0.1), dimensions={}, metadata={op_name="jit(broadcast_in_dim)/jit(main)/broadcast_in_dim[shape=(100, 100) broadcast_dimensions=()]" source_file="/var/folders/j_/lb9q5v61297652sxjndcrpq00000gn/T/ipykernel_1292/2814340469.py" source_line=3}

Failed after pipeline-start
```

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.28
jaxlib: 0.4.28
numpy:  1.26.4
python: 3.10.11 | packaged by conda-forge | (main, May 10 2023, 19:01:19) [Clang 14.0.6 ]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
platform: uname_result(system='Darwin', node='Maxims-MacBook-Air.local', release='22.1.0', version='Darwin Kernel Version 22.1.0: Sun Oct  9 20:14:30 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T8103', machine='arm64')
```

**Comment by jakevdp** [comment]
Hi - `int4` is experimental and has very little support across JAX. If you want to create an int4 array, one way to do so is like this:
```python
a = np.zeros((100, 100)).astype(np.int4)
```

**Comment by rajasekharporeddy** [comment]
Hi @Maxon081102 

The issue appears to have been resolved in JAX version 0.4.29. I've tested the code on a MacBook with an M1 Pro chip using JAX 0.4.29 and later versions, and it now runs without errors. Please see the attached screenshot for confirmation.

<img width="1493" alt="image" src="https://github.com/user-attachments/assets/dc61813d-56c2-4a50-8013-2f3fa0255df8">

Could you please upgrade to latest JAX version and check if the issue still persists?

Thank you.

**Comment by jakevdp** [comment]
Thanks for following up!
